ksp_plugin_test/plugin_compatibility_test.cpp

```diff
@@ -1,111 +1,134 @@
-using base::Array;
-using base::UniqueArray;
-using base::HexadecimalEncoder;
-using geometry::Bivector;
-using geometry::Trivector;
-using geometry::Vector;
-using quantities::Length;
-using quantities::si::Hour;
-using quantities::si::Metre;
-using quantities::si::Radian;
-using quantities::si::Second;
-using ::testing::AllOf;
-using ::testing::AnyOf;
-using ::testing::Gt;
-using ::testing::Lt;
-
-class TestablePlugin : public Plugin {
- public:
-  void KeepAllVessels();
-
-  std::map<GUID, not_null<Vessel const*>> vessels() const;
-
-  static not_null<std::unique_ptr<TestablePlugin>> ReadFromMessage(
-    serialization::Plugin const& message);
-};
+namespace interface {
 
-void TestablePlugin::KeepAllVessels() {
-  for (auto const& [_, vessel] : vessels_) {
-    kept_vessels_.insert(vessel.get());
-  }
-}
+using base::not_null;
+using base::OFStream;
+using base::PullSerializer;
+using base::PushDeserializer;
+using ksp_plugin::Plugin;
+using testing_utilities::ReadLinesFromBase64File;
+using testing_utilities::ReadLinesFromHexadecimalFile;
+using ::testing::NotNull;
 
-std::map<GUID, not_null<Vessel const*>> TestablePlugin::vessels() const {
-  std::map<GUID, not_null<Vessel const*>> result;
-  for (auto const& [guid, vessel] : vessels_) {
-    result.insert(std::make_pair(guid, vessel.get()));
-  }
-  return result;
-}
-
-not_null<std::unique_ptr<TestablePlugin>> TestablePlugin::ReadFromMessage(
-  serialization::Plugin const& message) {
-  std::unique_ptr<Plugin> plugin = Plugin::ReadFromMessage(message);
-  return std::unique_ptr<TestablePlugin>(
-      static_cast<TestablePlugin*>(plugin.release()));
-}
+const char preferred_compressor[] = "gipfeli";
+const char preferred_encoder[] = "base64";
 
 class PluginCompatibilityTest : public testing::Test {
  protected:
-  serialization::Plugin ReadFromFile(std::string const& filename) {
-    // Open the file and read hexadecimal data.
-    std::fstream file =
-        std::fstream(SOLUTION_DIR / "ksp_plugin_test" / filename);
-    CHECK(file.good());
-    std::string hex;
-    while (!file.eof()) {
-      std::string line;
-      std::getline(file, line);
-      for (auto const c : line) {
-        if ((c >= '0' && c <= '9') || (c >= 'A' && c <= 'F')) {
-          hex.append(1, c);
-        }
+  // Reads a plugin from a file containing only the "serialized_plugin = "
+  // lines, with "serialized_plugin = " dropped.
+  not_null<std::unique_ptr<Plugin const>> ReadPluginFromFile(
+      std::filesystem::path const& filename,
+      std::string_view const compressor,
+      std::string_view const encoder) {
+    Plugin const* plugin = nullptr;
+
+    PushDeserializer* deserializer = nullptr;
+    auto const lines =
+        encoder == "hexadecimal" ? ReadLinesFromHexadecimalFile(filename)
+        : encoder == "base64"    ? ReadLinesFromBase64File(filename)
+                                  : std::vector<std::string>{};
+    CHECK(!lines.empty());
+
+    LOG(ERROR) << "Deserialization starting";
```

Comment:
LOG(INFO) and google::LogToStdErr() in the constructor.

Reply:
Nope.  When I do that the output gets swamped with messages emitted from various places within the plugin (and its subobjects).  I want to be able to the progress of the test, not random crud unrelated to the test.